src/v/archival/async_data_uploader.h

```diff
@@ -0,0 +1,244 @@
+        return _params.value();
+    }
+
+    static ss::future<result<std::unique_ptr<segment_upload>>>
+    make_segment_upload(
+      ss::lw_shared_ptr<cluster::partition> part,
+      inclusive_offset_range range,
+      size_t read_buffer_size,
+      ss::scheduling_group sg,
+      model::timeout_clock::time_point deadline);
+
+    static ss::future<result<std::unique_ptr<segment_upload>>>
+    make_segment_upload(
+      ss::lw_shared_ptr<cluster::partition> part,
+      size_limited_offset_range range,
+      size_t read_buffer_size,
+      ss::scheduling_group sg,
+      model::timeout_clock::time_point deadline);
+
+private:
+    explicit segment_upload(
+      ss::lw_shared_ptr<cluster::partition> part,
+      size_t read_buffer_size,
+      ss::scheduling_group sg);
+
+    /// Initialize segment upload using offset range
+    ss::future<result<void>> initialize(
+      inclusive_offset_range offsets,
+      model::timeout_clock::time_point deadline);
+
+    void throw_if_not_initialized(std::string_view caller) const;
+
+    /// Stores parameters of the target batch.
+    /// The subscan returns this struct that contains
+    /// base/last offsets of the target batch.
+    struct subscan_res {
+        size_t filepos{0};
+        size_t num_batches{0};
+        size_t num_records{0};
+        model::timestamp first_timestamp;
+        model::timestamp max_timestamp;
+        model::offset base_offset;
+        model::offset last_offset;
+        bool done{false};
+    };
+
+    /// Scan partition to find target offset.
+    /// Returned 'subscan_res' struct contains size of the
+    /// scanned region on disk, number of batches/records,
+    /// the timestamp and offset of the 'target'.
+    /// The timestamp is a data timestamp (if the target is a config batch
+    /// the subscan will scan until the data batch is found and use first data
+    /// timestamp). The offset returned by the query is target or first offset >
+    /// target if target was compacted.
+    ss::future<result<subscan_res>> subscan(
+      ss::lw_shared_ptr<cluster::partition> part,
+      inclusive_offset_range range,
+      size_t initial_offset,
+      bool target_is_last_offset,
+      model::timeout_clock::time_point deadline) noexcept;
+
+    /// Calculate upload size using segment indexes
+    ss::future<result<cloud_upload_parameters>> compute_upload_parameters(
+      inclusive_offset_range range, model::timeout_clock::time_point deadline);
+
+    // Scan te beginning of the segment range
+    ss::future<result<subscan_res>> subscan_left(
+      std::vector<ss::lw_shared_ptr<storage::segment>> segments,
+      model::offset range_base,
+      model::timeout_clock::time_point deadline);
+
+    /// Scan the end of the segment range
+    ss::future<result<subscan_res>> subscan_right(
+      std::vector<ss::lw_shared_ptr<storage::segment>> segments,
+      model::offset range_last,
+      model::timeout_clock::time_point deadline);
```

Comment:
We have to upload an offset range that may span multiple segments and it may also start and stop in the middle of the segment. We have to find an upload candidate based on start offset and end offset of the upload or based on start offset and desired size of the upload. The algorithm is

- call `subscan_left` to find the exact beginning of the upload.
- call `subscan_right` to find the end of the upload.
- compute size and metadata of the upload based on two subscan results and segments in the middle.

The subscan uses segment index to find index entry closest to the target and then it scans up to the target to calculate precise file location. It may also scan past the target to find the data timestamp.